# Importing libraries

In [2]:
import numpy as np
import pandas as pd
import os


# Importing datasets

In [3]:
# Setting the path for the datasets
path = r'C:\Users\cpaul\Documents\Career Foundry\Instacart Basket Analysis'

# Importing datasets, orders and products
df_orders = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)
df_products = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_orders.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [6]:
df_orders['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [7]:
df_orders.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [8]:
df_orders.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [9]:
df_orders['order_id'] = df_orders['order_id'].astype('str')

In [10]:
df_orders['order_id'].dtype

dtype('O')

In [12]:
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [15]:
df_dep_t = df_dep.T

In [16]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [20]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [18]:
new_header = df_dep_t.iloc[0]

In [19]:
new_header

0    department
Name: department_id, dtype: object

In [21]:
df_dep_t_new = df_dep_t[1:]

In [22]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [23]:
df_dep_t_new.columns = new_header

In [24]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# Turning departments dataframe into a data dictionary

In [26]:
data_dict = df_dep_t_new.to_dict('index')

In [27]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [28]:
df_products.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [29]:
print(data_dict.get('19'))

{'department': 'snacks'}


# Subsetting 

In [30]:
df_products['department_id'] == 19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [32]:
df_products[df_products['department_id'] == 19]


product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [33]:
df_snacks = df_products[df_products['department_id'] == 19]

In [34]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [35]:
df_snacks_2 = df_products.loc[df_products['department_id'] == 19]

In [36]:
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [37]:
df_snacks_3 = df_products.loc[df_products['department_id'].isin([19])]

In [44]:
df_snacks_3.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# STEP 2: Changing variable in the df_orders dataframe that doesn't need to be included in the analysis as a numeric variable

In [39]:
# Viewing datatypes of df_orders dataframe 
df_orders.dtypes

order_id                   object
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [40]:
#Changing the datatype of the variable user_id
df_orders['user_id'] = df_orders['user_id'].astype('str')

In [42]:
df_orders['user_id'].dtype

dtype('O')

# STEP 3:Rename a variable with unintuitive name in df_orders dataframe without overwriting it

In [43]:
# Viewing columns of the dataframe df_orders
df_orders.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [45]:
# Renaming the variable days_since_prior_order
df_orders.rename(columns={'days_since_prior_order':'days_since_previous_order'}, inplace = True)

In [46]:
df_orders.head()

order_id user_id eval_set  order_number  orders_day_of_week  \
0  2539329       1    prior             1                   2   
1  2398795       1    prior             2                   3   
2   473747       1    prior             3                   3   
3  2254736       1    prior             4                   4   
4   431534       1    prior             5                   4   

   order_hour_of_day  days_since_previous_order  
0                  8                        NaN  
1                  7                       15.0  
2                 12                       21.0  
3                  7                       29.0  
4                 15                       28.0

# STEP 4: Busiest hour for placing orders

In [49]:
# Calling value_counts without the dropna = False arguement to determine frequency
df_orders['order_hour_of_day'].value_counts()

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

The busiest hour for placing orders is 10 AM, with a total order count of 288418.

# STEP 5: Determine the meaning behind the value of 4 in the department_id 

In [50]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [ ]:
# Using the get function 

In [51]:
print(data_dict.get('4'))

{'department': 'produce'}


Department Id 4 is produce.

# STEP 6: Create a subset with breakfast item sales

In [54]:
# Creating subset

df_breakfast = df_products[df_products['department_id']==14]

In [55]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

In [56]:
df_breakfast.shape

(1116, 5)

In [ ]:
There are 1116 products within the breakfast department.

# STEP 7: Creating subset of products to throw dinner parties - alcohol, deli, beverages, meat/seafood

In [57]:
df_dinner_parties = df_products.loc[df_products['department_id'].isin([5, 20, 7, 12])]

In [58]:
df_dinner_parties.head(20)

product_id                                       product_name  aisle_id  \
2            3               Robust Golden Unsweetened Oolong Tea        94   
6            7                     Pure Coconut Water With Orange        98   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
16          17                                  Rendered Duck Fat        35   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22          23                             Organic Turkey Burgers        49   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38          39           Daily Tangerine Citrus Flavored Beverage        64   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   
43          44                        Sparkling Raspberry Seltzer       115   
48          49      Vegetarian Grain Meat Sausages Italian - 4 CT        14   
51          52                                Mirabelle Brut Rose       134   
52          53              Whole Leaf Pure Aloe With Lemon Juice        98   
61          62            Premium Deli Oven Roasted Turkey Breast        96   
68          69                                                NaN        26   
72          73                  Jasmine Tea Unfiltered Ginger Ale        77   
73          74                    Artisan Chick'n & Apple Sausage        14   
81          82                       All Natural 100% Apple Juice        98   
83          84                                         Lamb Shank         7   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1  
19              7     6.0  
22             12     8.2  
34             12    15.1  
38              7    12.5  
39             12    22.5  
43              7    12.2  
48             20    10.1  
51              5    14.4  
52              7     4.7  
61             20    14.6  
68              7    11.8  
72              7     8.3  
73             20     3.0  
81              7     6.4  
83             12    24.3

In [59]:
# checking number of rows in the dataframe
df_dinner_parties.shape[0]

7650

# STEP 8: Number of rows in the df_dinner_parties dataframe

There are 7650 products under the departments(alcohol, deli, beverages, meat/seafood) that customers might use to throw dinner parties.

# STEP 9: Extract all the information about user_id of '1'

In [64]:
#Creating subset of orders dataframe for user_id of '1'(user_id field converted to str)

df_userid_1 = df_orders.loc[df_orders['user_id'] == '1']

In [65]:
df_userid_1.shape

(11, 7)

In [66]:
df_userid_1

order_id user_id eval_set  order_number  orders_day_of_week  \
0   2539329       1    prior             1                   2   
1   2398795       1    prior             2                   3   
2    473747       1    prior             3                   3   
3   2254736       1    prior             4                   4   
4    431534       1    prior             5                   4   
5   3367565       1    prior             6                   2   
6    550135       1    prior             7                   1   
7   3108588       1    prior             8                   1   
8   2295261       1    prior             9                   1   
9   2550362       1    prior            10                   4   
10  1187899       1    train            11                   4   

    order_hour_of_day  days_since_previous_order  
0                   8                        NaN  
1                   7                       15.0  
2                  12                       21.0  
3                   7                       29.0  
4                  15                       28.0  
5                   7                       19.0  
6                   9                       20.0  
7                  14                       14.0  
8                  16                        0.0  
9                   8                       30.0  
10                  8                       14.0

In [67]:
# Checking df_orders dataframe since I noticed "eval_set" column in the df_userid_1 dataframe above
df_orders.head()

order_id user_id eval_set  order_number  orders_day_of_week  \
0  2539329       1    prior             1                   2   
1  2398795       1    prior             2                   3   
2   473747       1    prior             3                   3   
3  2254736       1    prior             4                   4   
4   431534       1    prior             5                   4   

   order_hour_of_day  days_since_previous_order  
0                  8                        NaN  
1                  7                       15.0  
2                 12                       21.0  
3                  7                       29.0  
4                 15                       28.0

In [68]:
# Dropping the eval_set column in df_orders dataframe, not sure how it got back in, I had dropped the column before.
# Noticed the column while displaying data for df_userid_1 dataframe
df_orders.drop(columns = ['eval_set'])

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
1        2398795       1             2                   3                  7   
2         473747       1             3                   3                 12   
3        2254736       1             4                   4                  7   
4         431534       1             5                   4                 15   
...          ...     ...           ...                 ...                ...   
3421078  2266710  206209            10                   5                 18   
3421079  1854736  206209            11                   4                 10   
3421080   626363  206209            12                   1                 12   
3421081  2977660  206209            13                   1                 12   
3421082   272231  206209            14                   6                 14   

         days_since_previous_order  
0                              NaN  
1                             15.0  
2                             21.0  
3                             29.0  
4                             28.0  
...                            ...  
3421078                       29.0  
3421079                       30.0  
3421080                       18.0  
3421081                        7.0  
3421082                       30.0  

[3421083 rows x 6 columns]

In [69]:
# Dropping the eval_set column on the df_userid_1 dataframe also
df_userid_1.drop(columns = ['eval_set'])

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_previous_order  
0                         NaN  
1                        15.0  
2                        21.0  
3                        29.0  
4                        28.0  
5                        19.0  
6                        20.0  
7                        14.0  
8                         0.0  
9                        30.0  
10                       14.0

In [ ]:
There are 11 orders from user_id '1'. The orders were made on days 1,2,3,4 of the week.

# STEP 10: Basis stats about the user_id '1'

In [70]:
# Summary statistics of df_userid_1
df_userid_1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_previous_order  
count                  10.000000  
mean                   19.000000  
std                     9.030811  
min                     0.000000  
25%                    14.250000  
50%                    19.500000  
75%                    26.250000  
max                    30.000000

# STEP 12: Exporting df_orders dataframe

In [71]:
#Exporting df_orders dataframe to "Prepared Data" folder
df_orders.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'))

# STEP 13: Exporting df_dep_t_new dataframe

In [72]:
#Exporting df_dep_t_new dataframe to "Prepared Data" folder
df_dep_t_new.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'departments_wrangled.csv'))